server/src/main/java/org/opensearch/common/cache/store/EhCacheDiskCache.java

```diff
@@ -0,0 +1,543 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.cache.store;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.OpenSearchException;
+import org.opensearch.common.SuppressForbidden;
+import org.opensearch.common.cache.LoadAwareCacheLoader;
+import org.opensearch.common.cache.RemovalReason;
+import org.opensearch.common.cache.stats.CacheStats;
+import org.opensearch.common.cache.store.builders.StoreAwareCacheBuilder;
+import org.opensearch.common.cache.store.enums.CacheStoreType;
+import org.opensearch.common.cache.store.listeners.StoreAwareCacheEventListener;
+import org.opensearch.common.collect.Tuple;
+import org.opensearch.common.metrics.CounterMetric;
+import org.opensearch.common.settings.Setting;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.unit.TimeValue;
+
+import java.io.File;
+import java.time.Duration;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.function.BiFunction;
+import java.util.function.Supplier;
+
+import org.ehcache.Cache;
+import org.ehcache.CachePersistenceException;
+import org.ehcache.PersistentCacheManager;
+import org.ehcache.config.builders.CacheConfigurationBuilder;
+import org.ehcache.config.builders.CacheEventListenerConfigurationBuilder;
+import org.ehcache.config.builders.CacheManagerBuilder;
+import org.ehcache.config.builders.PooledExecutionServiceConfigurationBuilder;
+import org.ehcache.config.builders.ResourcePoolsBuilder;
+import org.ehcache.config.units.MemoryUnit;
+import org.ehcache.event.CacheEvent;
+import org.ehcache.event.CacheEventListener;
+import org.ehcache.event.EventType;
+import org.ehcache.expiry.ExpiryPolicy;
+import org.ehcache.impl.config.store.disk.OffHeapDiskStoreConfiguration;
+import org.ehcache.spi.loaderwriter.CacheLoadingException;
+import org.ehcache.spi.loaderwriter.CacheWritingException;
+
+/**
+ * This variant of disk cache uses Ehcache underneath.
+ *  @param <K> Type of key.
+ *  @param <V> Type of value.
+ *
+ *  @opensearch.experimental
+ *
+ */
+public class EhCacheDiskCache<K, V> implements StoreAwareCache<K, V> {
```

Comment:
When operating a production service, I like smaller cycles to tune features and see performances tradeoffs.  Can this cache be enabled/disabled on an index handling live requests - if not what would it take to get there?

This aligns with @msfroh [[comment]](https://github.com/opensearch-project/OpenSearch/issues/10024#issuecomment-1730173944) about how synthetic tests are game-able whereas for OpenSearch operators we want to enable them to game their specific scenarios' performance.  What do you think about supporting live tweaking of these caches?

Reply:
@peternied Yes, we will have that. We will have a setting(feature flag as a start) which defines a cache type setting(onHeap or tiered) but requires a node restart.
Within tiered(onHeap + disk) cache type as well, we will be exposing a setting to turn off disk dynamically(no spillover), so that it essentially behaves as onHeap cache.